internal/gps/pkgtree/digest.go

```diff
@@ -0,0 +1,442 @@
+// Copyright 2017 The Go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package pkgtree
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"fmt"
+	"hash"
+	"io"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	pathSeparator = string(filepath.Separator)
+
+	// when walking vendor root hierarchy, ignore file system nodes of the
+	// following types.
+	skipSpecialNodes = os.ModeDevice | os.ModeNamedPipe | os.ModeSocket | os.ModeCharDevice
+)
+
+// DigestFromPathname returns a deterministic hash of the specified file system
+// node, performing a breadth-first traversal of directories. While the
+// specified prefix is joined with the pathname to walk the file system, the
+// prefix string is eliminated from the pathname of the nodes encounted when
+// hashing the pathnames, so that the resultant hash is agnostic to the absolute
+// root directory path of the nodes being checked.
+//
+// This function ignores any file system node named `vendor`, `.bzr`, `.git`,
+// `.hg`, and `.svn`, as these are typically used as Version Control System
+// (VCS) directories.
+//
+// Other than the `vendor` and VCS directories mentioned above, the calculated
+// hash includes the pathname to every discovered file system node, whether it
+// is an empty directory, a non-empty directory, empty file, non-empty file, or
+// symbolic link. If a symbolic link, the referent name is included. If a
+// non-empty file, the file's contents are incuded. If a non-empty directory,
+// the contents of the directory are included.
+//
+// While filepath.Walk could have been used, that standard library function
+// skips symbolic links, and for now, we want the hash to include the symbolic
+// link referents.
+func DigestFromPathname(prefix, pathname string) ([]byte, error) {
```

Comment:
i'm trying to figure out the best way to communicate this, because i roughly see the end use pattern you're orienting towards, and as long as you're orienting toward that, i think we may have communication disconnects due to these mismatched assumptions. i've tried a couple times, but it seems it hasn't quite landed (which is fine, this is complex), so i'm not sure what to say, apart from...maybe try to take a step back and be a little more meta in interpreting what i'm saying?

> Also keep in mind, as you reminded me, that there is no standard number of directories down from the vendor root to the project location.

right, and i'm not suggesting any magic discovery here. i'm saying that, to reuse your example, `VerifyDepTree()` _should_ call `DigestFromPathname()` with `"../some/path/to/vendor/github.com/alice/alice1"` because we do _not_ want `github.com/alice/alice1` - the import path portion of the prefix - to be included in the hash. it is a _good_ thing that the hash would be the same independent of this prefix. 

consider `gopkg.in` - it is a potentially useful thing that a tree from `gopkg.in/yaml.v2` and `github.com/go-yaml/yaml` would have the same hash. they ultimately resolve to the same repository, and choosing a hashing scheme that discounts the root portion would allow a content-addressable system based on this hash to avoid duplicating storage for packages built from the ostensibly different import paths.

basically, i'm saying that we're doing a bit too much here, and `DigestFromPathname()` can be dumber. its responsibility can safely be reduced to just stupidly hashing the tree we point it at, and the question of how we organize that vis-a-vis the import root can be left to the caller.